### DISCUSSION TITLE
airflow scheduler container often hangs up without restarting results in no heartbeat issue

### DISCUSSION THREAD

**Discussion by lazyprogammer** [apache/airflow#33200] [General]
I'm running airflow in Kubernetes (Azure)
Meta data DB is Azure PaaS Postgres version 11
airflow version is 2.2.5
airflow scheduler is running with 3 replicas

Below are the common errors occuring after which the scheduler gets hanged up causing scheduler no hearbeat issue and all the scheduler jobs are running as zombie process

    "Can't reconnect until invalid "
sqlalchemy.exc.InvalidRequestError: Can't reconnect until invalid transaction is rolled back

sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server conn crashed?
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/default.py:410 SAWarning: Exception attempting to detect unicode returns: OperationalError('(psycopg2.OperationalError) server conn crashed?\nserver closed the connection unexpectedly\n\tThis probably means the server terminated abnormally\n\tbefore or while processing the request.\n')

**Comment by potiuk** [comment]
Likely you should work on stability of your Airflow -> Postgres comnnections. 

Or Maybe simply your postgres is crashing or your connection is instable. Or your postgres lacks resorurces. 

Or maybe you have not followed the best practices for production deployment and:

1) You do not use pgbouncer (see notes in https://airflow.apache.org/docs/apache-airflow/stable/howto/set-up-database.html#setting-up-a-postgresql-database) 
2) Yoy have a lot of Variable and other DB operation in Top-level code of your DAGs https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code 

If those two are combined (and even just 1, you do not have pgbouncer) - airflow might open a lot of connection to Postgres and each connection in Postgres is a separate process, so that might lead to your Postgres exhausting resources.

In any event if there is any reason why your postgres is closing connection, you should look for the reason in the postgres logs. They will tell you what's going on
